core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -582,6 +723,16 @@ public ShardRouting moveToUnassigned(ShardRouting shard, UnassignedInfo unassign
         return unassigned;
     }
 
+    private ShardRouting movePrimaryToUnassignedAndDemoteToReplica(ShardRouting shard, UnassignedInfo unassignedInfo) {
+        ensureMutable();
```

Comment:
I've reworked the `ensureMutable();` checks so that they are only done on the public methods that modify the routing nodes.